test/e2e/clusterctl_upgrade.go

```diff
@@ -388,6 +391,11 @@ func ClusterctlUpgradeSpec(ctx context.Context, inputGetter func() ClusterctlUpg
 			kubernetesVersion = input.E2EConfig.GetVariable(KubernetesVersion)
 		}
 		controlPlaneMachineCount := ptr.To[int64](1)
+
+		// Don't set controlPlaneMachineCount to 0 to satisfy the validation in the cluster template.
+		if input.ControlPlaneMachineCount != nil && input.ControlPlaneMachineCount != ptr.Int64(0) {
+			controlPlaneMachineCount = input.ControlPlaneMachineCount
+		}
```

Comment:
This has to be done because the [call to `clusterctl.ConfigClusterWithBinary`](https://github.com/willie-yao/cluster-api/blob/clusterctl-upgrade-cpcount/test/e2e/clusterctl_upgrade.go#L405) does not allow `ControlPlaneMachineCount` to be set to 0. The parameter for `ControlPlaneMachineCount` does not affect the cluster template for scenarios such as AKS Managed Clusters.

Reply:
I would prefer to extend ConfigClusterWithBinary so it can handle 0